src/ompl/base/spaces/Dubins3DMotionValidator.h

```diff
@@ -0,0 +1,179 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Mark Moll */
+
+#ifndef OMPL_BASE_SPACES_DUBINS3D_MOTION_VALIDATOR_
+#define OMPL_BASE_SPACES_DUBINS3D_MOTION_VALIDATOR_
+
+#include "ompl/base/SpaceInformation.h"
+#include <queue>
+
+namespace ompl::base
+{
+
+    /**
+     * \brief A 3D Dubins plane motion validator that only uses the state validity checker.
+     *  Motions are checked for validity at a specified resolution.
+     *
+     * This motion validator is almost identical to the DiscreteMotionValidator except that it remembers the optimal
+     * PathType between different calls to interpolate. It is intended to be used with the OwenStateSpace,
+     * VanaStateSpace, and VanaOwenStateSpace. The SpaceInformation::setMotionValidator method uses this as the default
+     * for those state spaces.
+     */
+    template <class Dubins3DStateSpace>
+    class Dubins3DMotionValidator : public MotionValidator
+    {
+    public:
+        Dubins3DMotionValidator(SpaceInformation *si) : MotionValidator(si)
+        {
+            defaultSettings();
+        }
+        Dubins3DMotionValidator(const SpaceInformationPtr &si) : MotionValidator(si)
+        {
+            defaultSettings();
+        }
+        ~Dubins3DMotionValidator() override = default;
+        bool checkMotion(const State *s1, const State *s2) const override
+        {
+            /* assume motion starts in a valid configuration so s1 is valid */
+            if (!si_->isValid(s2))
+                return false;
+            auto path = stateSpace_->getPath(s1, s2);
+            if (!path)
+                return false;
+
+            bool result = true;
+            int nd = stateSpace_->validSegmentCount(s1, s2);
+
+            /* initialize the queue of test positions */
+            std::queue<std::pair<int, int>> pos;
+            if (nd >= 2)
+            {
+                pos.emplace(1, nd - 1);
+
+                /* temporary storage for the checked state */
+                State *test = si_->allocState();
```

Comment:
Can this be moved to the ctor/dtor? So that allocation is not part of the planner sampling loop

Reply:
that would violate thread safety